From: Hannes Reinecke <hare@suse.de>
Date: Tue, 9 May 2017 13:25:04 +0200
Subject: scsi: avoid system stall due to host_busy race
References: bsc#1031358
Patch-Mainline: never, SLE11 specific

There is a race condition in the SCSI stack which results in
the 'host_busy' value to decrement below zero.
This patch will avoid a system stall for these cases and put out
a warning if the race occurs.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/scsi/scsi_lib.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/scsi/scsi_lib.c b/drivers/scsi/scsi_lib.c
index c61dbe4..4514557 100644
--- a/drivers/scsi/scsi_lib.c
+++ b/drivers/scsi/scsi_lib.c
@@ -304,8 +304,14 @@ void scsi_device_unbusy(struct scsi_device *sdev)
 	unsigned long flags;
 
 	spin_lock_irqsave(shost->host_lock, flags);
-	shost->host_busy--;
-	starget->target_busy--;
+	if (unlikely(shost->host_busy == 0))
+		sdev_printk(KERN_WARNING, sdev, "host_busy already 0\n");
+	else
+		shost->host_busy--;
+	if (unlikely(starget->target_busy == 0))
+		sdev_printk(KERN_WARNING, sdev, "target_busy already 0\n");
+	else
+		starget->target_busy--;
 	if (unlikely(scsi_host_in_recovery(shost) &&
 		     (shost->host_failed || shost->host_eh_scheduled)))
 		scsi_eh_wakeup(shost);
-- 
1.8.5.6

